File: python/src/adapters/email/ouraihub.py
"""OurAIHub temporary email client (mail.ouraihub.com)."""

from __future__ import annotations

from dataclasses import dataclass

import httpx

from adapters.email.base import EmailClient, EmailMessage

BASE_URL = 'https://mail.ouraihub.com/api'


@dataclass(frozen=True)
class OuraihubConfig:
    api_key: str
    domain: str = 'ouraihub.com'
    expiry_time: int = 3600000  # 1 hour


class OuraihubClient(EmailClient):
    """Temporary email via mail.ouraihub.com API.

    Usage:
        client = OuraihubClient(OuraihubConfig(api_key='mk_xxx'))
        client.create_email('myprefix')  # creates myprefix@moemail.app
        messages = client.fetch_recent_messages()
    """

    def __init__(self, config: OuraihubConfig):
        self._config = config
        self._http = httpx.Client(timeout=30)
        self._email_id: str | None = None
        self._email_address: str | None = None

    @property
    def headers(self) -> dict[str, str]:
        return {'X-API-Key': self._config.api_key}

    @property
    def email_address(self) -> str | None:
        return self._email_address

    def get_available_domains(self) -> list[str]:
        """Get available email domains from system config."""
        resp = self._http.get(f'{BASE_URL}/config', headers=self.headers)
        if resp.status_code != 200:
            return [self._config.domain]
        data = resp.json()
        # Try common response shapes
        if isinstance(data, dict):
            domains = data.get('domains') or data.get('data', {}).get('domains')
            if isinstance(domains, list):
                return domains
        return [self._config.domain]

    def create_email(self, name: str) -> str | None:
        """Generate a temporary email address. Returns the full email address."""
        resp = self._http.post(
            f'{BASE_URL}/emails/generate',
            headers={**self.headers, 'Content-Type': 'application/json'},
            json={
                'name': name,
                'expiryTime': self._config.expiry_time,
                'domain': self._config.domain,
            },
        )
        if resp.status_code not in (200, 201):
            return None
        data = resp.json()
        if isinstance(data, dict):
            self._email_id = data.get('id')
            self._email_address = data.get('email') or data.get('address')
            if not self._email_address and self._email_id:
                self._email_address = f'{name}@{self._config.domain}'
        return self._email_address

    def set_email_id(self, email_id: str) -> None:
        """Set email ID directly if already created."""
        self._email_id = email_id

    def fetch_recent_messages(self, limit: int = 50) -> list[EmailMessage]:
        if not self._email_id:
            return []
        resp = self._http.get(
            f'{BASE_URL}/emails/{self._email_id}',
            headers=self.headers,
        )
        if resp.status_code != 200:
            return []
        data = resp.json()

        items = data.get('messages', []) if isinstance(data, dict) else []
        messages: list[EmailMessage] = []
        for item in items[:limit]:
            body = item.get('html') or item.get('content') or item.get('text') or ''
            messages.append(EmailMessage(
                id=str(item.get('id', '')),
                sender=item.get('from_address') or item.get('from') or '',
                subject=str(item.get('subject', '')),
                body_text=body,
                received_at=str(item.get('received_at') or item.get('sent_at') or ''),
            ))
        return messages

File: node-register/src/email/ouraihub.ts
import { TempEmailService, extractCode } from './email-service'

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms))
}

export class OuraihubEmailService implements TempEmailService {
  private apiKey: string
  private domain: string
  private address = ''
  private emailId = ''
  private baseURL = 'https://mail.ouraihub.com/api'

  constructor(apiKey: string, domain: string) {
    this.apiKey = apiKey
    this.domain = domain
  }

  getAddress(): string {
    return this.address
  }

  async create(): Promise<string> {
    const name = 'kiro' + Date.now().toString(36) + Math.random().toString(36).slice(2, 6)

    const resp = await fetch(`${this.baseURL}/emails/generate`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', 'X-API-Key': this.apiKey },
      body: JSON.stringify({ name, expiryTime: 3600000, domain: this.domain })
    })
    if (!resp.ok) {
      const text = await resp.text()
      throw new Error(`OurAIHub create email failed (${resp.status}): ${text}`)
    }
    const data = await resp.json() as { id?: string; email?: string; address?: string }
    this.emailId = data.id || ''
    this.address = data.email || data.address || `${name}@${this.domain}`
    return this.address
  }

  async waitForCode(timeoutSec: number, intervalSec: number): Promise<string> {
    if (!this.emailId) throw new Error('emailId 未设置，请先调用 create()')
    const deadline = Date.now() + timeoutSec * 1000
    while (Date.now() < deadline) {
      const resp = await fetch(`${this.baseURL}/emails/${this.emailId}`, {
        headers: { 'X-API-Key': this.apiKey }
      })
      if (resp.ok) {
        const data = await resp.json() as { messages?: Array<{ html?: string; content?: string; text?: string }> }
        if (data.messages && data.messages.length > 0) {
          for (const msg of data.messages) {
            const code = extractCode(msg.html || msg.content || msg.text || '')
            if (code) return code
          }
        }
      }
      await sleep(intervalSec * 1000)
    }
    throw new Error(`等待验证码超时 (${timeoutSec}s)`)
  }
}
